app/lib/backup/runner.ts

```diff
@@ -22,154 +26,215 @@ export interface Options {
    * not yet stored.
    */
   onDialogRetrieved?: (id: bigint) => unknown
-  /**
-   * Called when an entire dialog has been stored with Storacha.
-   */
-  onDialogStored?: (id: bigint) => unknown
 }
 
 export const run = async (ctx: Context, space: SpaceDID, dialogs: Set<bigint>, period: AbsolutePeriod, options?: Options): Promise<UnknownLink> => {
-  const { onDialogStored, onDialogRetrieved } = options ?? {}
-  const dialogMessages: Record<string, UnknownLink> = {}
+  const dialogDatas: BackupData['dialogs'] = {}
+  const pendingDialogIDs = Array.from(dialogs)
+
+  // null value signals that the current dialog has completed
+  let dialogEntity: Entity | null = null
+
+  let entities: EntityRecordData = {}
+  let messages: MessageData[] = []
+  let messageLinks: Array<Link<EncryptedByteView<MessageData[]>>> = []
+
+  // null value signals that no more messages will come and the current dialog
+  // can now be finalized
+  let messageIterator: AsyncIterator<Api.Message> | null = null
+
+  const blockStream = new ReadableStream<Block>({
+    async pull (controller) {
+      if (!dialogEntity) {
+        // get the next dialog to back up
+        const dialogID = pendingDialogIDs.shift()
+        if (!dialogID) {
+          // we finished!
+          console.log(`creating root with ${Object.entries(dialogDatas).length} dialogs`)
+          const rootData: BackupModel = {
+            [versionTag]: { dialogs: dialogDatas, period }
+          }
+          const rootBlock = await encodeBlock({ value: rootData, codec: dagCBOR, hasher: sha256 })
+          controller.enqueue(rootBlock)
+          controller.close()
+          return
+        }
 
-  console.log('chats: ', dialogs)
-  const selectedChats = Array.from(dialogs)
+        dialogEntity = await ctx.telegram.getEntity(dialogID)
+        const [firstMessage] = await ctx.telegram.getMessages(dialogID, {
+          limit: 1,
+          // FML, there is a falsey check on this value so we cannot pass 0
```

Comment:
But we can pass `undefined`, which will make it ignore the date and just fetch the latest message. In cases where we don't have a `from` date, we can simply use `iterMessages` without specifying `offsetDate` or `minId`

Reply:
Ok so here we're trying to get the first message that's older than our start date right?

The problem is that if you pass `0` you get the newest message, not an old message and not "no message". Really when we pass `0` we want no message to come back so that `minId` in the call to `iterMessages` will be `undefined`.

I'm going to refactor to avoid the call to `getMessages` altogether when `period[0] === 0` since we're not expecting anything to come back.